mongo/integration/unified/testrunner_operation.go

```diff
@@ -111,11 +121,96 @@ func executeTestRunnerOperation(ctx context.Context, operation *operation) error
 		coll := lookupString(args, "collectionName")
 		index := lookupString(args, "indexName")
 		return verifyIndexExists(ctx, db, coll, index, false)
+	case "loop":
+		unmarshaledArgs := &loopArgs{}
+		if err := bson.Unmarshal(args, unmarshaledArgs); err != nil {
+			return err
+		}
+		return loop(ctx, unmarshaledArgs, r)
 	default:
 		return fmt.Errorf("unrecognized testRunner operation %q", operation.Name)
 	}
 }
 
+func loop(ctx context.Context, args *loopArgs, r *Runner) error {
+	if r == nil {
+		return fmt.Errorf("Runner should be set for looping operation")
+	}
+
+	// setup entities
+	entityMap := entities(ctx)
+	if args.StoreErrors != "" {
+		if err := entityMap.addErrorsEntityIfDoesntExist(args.StoreErrors); err != nil {
+			return err
+		}
+	}
+	if args.StoreFailures != "" {
+		if err := entityMap.addFailuresEntityIfDoesntExist(args.StoreFailures); err != nil {
+			return err
+		}
+	}
+	if args.StoreSuccesses != "" {
+		if err := entityMap.addSuccessesEntity(args.StoreSuccesses); err != nil {
+			return err
+		}
+	}
+	if args.StoreIterations != "" {
+		if err := entityMap.addIterationsEntity(args.StoreIterations); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-r.loopDone:
+			return nil
+		default:
+			var loopErr error
+			for _, operation := range args.Operations {
+				if operation.Name == "loop" {
+					return fmt.Errorf("Loop sub-operations should not include loop")
+				}
+				loopErr = operation.execute(ctx, r)
+
+				// if the operation errors, stop this loop
+				if loopErr != nil {
+					if args.StoreErrors == "" && args.StoreFailures == "" {
+						break
+					}
+					record := bson.D{
+						{"error", loopErr.Error()},
+						{"time", time.Now().Unix()},
+					}
+					errDoc, _ := bson.Marshal(record)
+					switch {
+					case args.StoreErrors == "":
+						entityMap.appendFailuresEntity(args.StoreFailures, errDoc)
+					case args.StoreFailures == "":
+						entityMap.appendErrorsEntity(args.StoreErrors, errDoc)
+					// errors are test runner errors
+					case strings.Contains(loopErr.Error(), "execution failed: "):
```

Comment:
I know we discussed string checking as an option and it works fine, but it's definitely not ideal. If you don't want to increase the scope of this PR, we should at least file a ticket to fix this using custom error types and a type switch.